### PR TITLE
TBOX-240: add util function to get recipe id

### PR DESCRIPTION
**Pull request by nguyentr17**
# ↪️ Pull Request

## ✔️ PR Todo

- [x] Add documentation
- [ ] Add examples
- [x] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [x] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [x] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [ ] Approval from first code-reviewer
- [ ] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [x] Pass all checks

### CODE REVIEW

**Comment by nguyentr17** [comment]
I plan to add an example to run high-impact cluster operation using the dedup recipe. Is this okay to include an example using unversioned API endpoints? @skalish @jun-dao @Kovacsova @ycharm @wpmartin

File: tamr_toolbox/utils/recipe.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        project: a mastering project
+
+    Returns: dedup recipe id of the project.
+
+    """
+    # Retrieve project steps using unified dataset
+    project_steps = project.unified_dataset().usage().usage.input_to_project_steps
+    # Search for dedup project step of the mastering project
+    for step in project_steps:
+        if step.project_name == project.name and step.type == ProjectType.DEDUP.value:
```
Comment (ycharm) [inline tamr_toolbox/utils/recipe.py:26]: This breaks if the project name is manually changed after creating the project, is there a reason for this part? 
```suggestion
        if step.type == ProjectType.DEDUP.value:
```

File: tamr_toolbox/utils/recipe.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    Retrieves dedup recipe ID of a mastering project.
+    To be used to run custom recipe operations.
+
+    Args:
+        project: a mastering project
+
+    Returns: dedup recipe id of the project.
+
+    """
+    # Retrieve project steps using unified dataset
```
Comment (ycharm) [inline tamr_toolbox/utils/recipe.py:22]: perhaps we could add a check here to make sure the type is consistent and raise a TypeError:
```suggestion

    #check project type
    not_mastering_error = (
        f"{project.name} is not a mastering project."
    )
    if project.type != ProjectType.DEDUP.value:
        LOGGER.error(not_mastering_error)
        raise Exception(not_mastering_error)
    # Retrieve project steps using unified dataset
```

**Comment by skalish** [comment]
@nguyentr17 I think we should revisit this feature.  I'm concerned that `recipe`s are not really well-supported in a versioned API Tamr pipeline, and are more of an internal detail.  While your code here is very good, I think implementing a `utils.recipe` module may need to wait for the product to better support interacting with them.  We should avoid treating `utils` as a miscellaneous bucket for features, since this will be hard to maintain over time.  Let me know if you want to discuss further.

**Comment by skalish** [comment]
The current plan is to create a knowledge base article around this, or fold it into another project, so we'll keep this as a draft for now.

**Comment by skalish** [comment]
Closing for now as we need to wait for versioned APIs to implement as part of `tamr-toolbox`.

**Pull request closed without merging**